File: app/services/steganography.py
from __future__ import annotations

import base64
import json
import os
from dataclasses import dataclass
from io import BytesIO
from pathlib import Path

from cryptography.exceptions import InvalidTag
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from PIL import Image, UnidentifiedImageError

MAGIC = b"SIHS1"
FORMAT_VERSION = 1
KDF_ITERATIONS = 390_000
SUPPORTED_IMAGE_FORMATS = {"PNG", "BMP"}
SUPPORTED_EXTENSIONS = {".png", ".bmp"}


class SteganographyError(ValueError):
    pass


class UnsupportedImageError(SteganographyError):
    pass


class CapacityError(SteganographyError):
    pass


class NoHiddenDataError(SteganographyError):
    pass


class PasswordRequiredError(SteganographyError):
    pass


class InvalidPasswordError(SteganographyError):
    pass


@dataclass(slots=True)
class HiddenPayload:
    kind: str
    data: bytes
    filename: str | None
    content_type: str
    encrypted: bool


def embed_secret_in_cover(
    cover_bytes: bytes,
    cover_filename: str,
    *,
    secret_data: bytes,
    payload_kind: str,
    payload_filename: str | None = None,
    content_type: str | None = None,
    password: str | None = None,
) -> bytes:
    package = build_package(
        secret_data=secret_data,
        payload_kind=payload_kind,
        payload_filename=payload_filename,
        content_type=content_type,
        password=password,
    )
    image = open_supported_image(cover_bytes, cover_filename)
    return embed_package(image, package)


def extract_secret_from_image(stego_bytes: bytes, password: str | None = None) -> HiddenPayload:
    image = open_supported_image(stego_bytes, "stego.png", check_extension=False)
    package = extract_package(image)
    return parse_package(package, password=password)


def capacity_bytes(cover_bytes: bytes, cover_filename: str) -> int:
    image = open_supported_image(cover_bytes, cover_filename)
    return max((image.width * image.height * 3) // 8 - 4, 0)


def open_supported_image(image_bytes: bytes, filename: str, *, check_extension: bool = True) -> Image.Image:
    if check_extension and Path(filename).suffix.lower() not in SUPPORTED_EXTENSIONS:
        raise UnsupportedImageError("Only PNG and BMP images are supported.")

    try:
        with Image.open(BytesIO(image_bytes)) as image:
            if image.format not in SUPPORTED_IMAGE_FORMATS:
                raise UnsupportedImageError("Only PNG and BMP images are supported.")
            image.load()
            return image.convert("RGB")
    except UnidentifiedImageError as exc:
        raise UnsupportedImageError("The uploaded file is not a readable image.") from exc


def build_package(
    *,
    secret_data: bytes,
    payload_kind: str,
    payload_filename: str | None,
    content_type: str | None,
    password: str | None,
) -> bytes:
    if payload_kind not in {"text", "file"}:
        raise SteganographyError("Secret payload must be text or file.")
    if not secret_data:
        raise SteganographyError("Secret payload cannot be empty.")

    encrypted = bool(password)
    payload_bytes = secret_data
    metadata: dict[str, object] = {
        "version": FORMAT_VERSION,
        "kind": payload_kind,
        "filename": payload_filename,
        "content_type": content_type or "application/octet-stream",
        "encrypted": encrypted,
    }

    if encrypted:
        salt = os.urandom(16)
        nonce = os.urandom(12)
        key = derive_key(password or "", salt, KDF_ITERATIONS)
        payload_bytes = AESGCM(key).encrypt(nonce, secret_data, MAGIC)
        metadata.update(
            {
                "kdf": {
                    "name": "PBKDF2HMAC-SHA256",
                    "iterations": KDF_ITERATIONS,
                    "salt": encode_b64(salt),
                },
                "nonce": encode_b64(nonce),
            }
        )

    header = json.dumps(metadata, separators=(",", ":"), sort_keys=True).encode("utf-8")
    return MAGIC + len(header).to_bytes(4, "big") + header + payload_bytes


def parse_package(package: bytes, *, password: str | None) -> HiddenPayload:
    if not package.startswith(MAGIC):
        raise NoHiddenDataError("No hidden SIHS payload was found.")
    if len(package) < len(MAGIC) + 4:
        raise NoHiddenDataError("The hidden payload is incomplete.")

    offset = len(MAGIC)
    header_length = int.from_bytes(package[offset : offset + 4], "big")
    header_start = offset + 4
    header_end = header_start + header_length
    if header_length <= 0 or header_end > len(package):
        raise NoHiddenDataError("The hidden payload metadata is invalid.")

    try:
        metadata = json.loads(package[header_start:header_end].decode("utf-8"))
    except (json.JSONDecodeError, UnicodeDecodeError) as exc:
        raise NoHiddenDataError("The hidden payload metadata is invalid.") from exc

    if metadata.get("version") != FORMAT_VERSION:
        raise NoHiddenDataError("The hidden payload version is not supported.")

    payload_bytes = package[header_end:]
    encrypted = bool(metadata.get("encrypted"))
    if encrypted:
        if not password:
            raise PasswordRequiredError("A password is required to extract this payload.")
        try:
            kdf = metadata["kdf"]
            salt = decode_b64(str(kdf["salt"]))
            iterations = int(kdf["iterations"])
            nonce = decode_b64(str(metadata["nonce"]))
            key = derive_key(password, salt, iterations)
            payload_bytes = AESGCM(key).decrypt(nonce, payload_bytes, MAGIC)
        except (KeyError, ValueError, TypeError, InvalidTag) as exc:
            raise InvalidPasswordError("The provided password is incorrect.") from exc

    kind = str(metadata.get("kind") or "")
    if kind not in {"text", "file"}:
        raise NoHiddenDataError("The hidden payload type is not supported.")

    filename = metadata.get("filename")
    return HiddenPayload(
        kind=kind,
        data=payload_bytes,
        filename=str(filename) if filename else None,
        content_type=str(metadata.get("content_type") or "application/octet-stream"),
        encrypted=encrypted,
    )


def embed_package(image: Image.Image, package: bytes) -> bytes:
    frame = len(package).to_bytes(4, "big") + package
    capacity_bits = image.width * image.height * 3
    required_bits = len(frame) * 8
    if required_bits > capacity_bits:
        raise CapacityError(
            f"Secret payload needs {required_bits // 8} bytes, but this image can hold {capacity_bits // 8} bytes."
        )

    channels = rgb_channels(image)
    for index, bit in enumerate(bytes_to_bits(frame)):
        channels[index] = (channels[index] & 0xFE) | bit

    pixels = [tuple(channels[index : index + 3]) for index in range(0, len(channels), 3)]
    output = Image.new("RGB", image.size)
    output.putdata(pixels)

    buffer = BytesIO()
    output.save(buffer, format="PNG")
    return buffer.getvalue()


def extract_package(image: Image.Image) -> bytes:
    channels = rgb_channels(image)
    if len(channels) < 32:
        raise NoHiddenDataError("No hidden SIHS payload was found.")

    package_length = int.from_bytes(bits_to_bytes((channel & 1 for channel in channels[:32]), 4), "big")
    capacity = len(channels) // 8 - 4
    if package_length <= 0 or package_length > capacity:
        raise NoHiddenDataError("No hidden SIHS payload was found.")

    start = 32
    end = start + package_length * 8
    package = bits_to_bytes((channel & 1 for channel in channels[start:end]), package_length)
    if not package.startswith(MAGIC):
        raise NoHiddenDataError("No hidden SIHS payload was found.")
    return package


def derive_key(password: str, salt: bytes, iterations: int) -> bytes:
    kdf = PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=32,
        salt=salt,
        iterations=iterations,
    )
    return kdf.derive(password.encode("utf-8"))


def rgb_channels(image: Image.Image) -> list[int]:
    if hasattr(image, "get_flattened_data"):
        pixels = image.get_flattened_data()
    else:
        pixels = image.getdata()
    return [channel for pixel in pixels for channel in pixel[:3]]


def bytes_to_bits(data: bytes):
    for byte in data:
        for shift in range(7, -1, -1):
            yield (byte >> shift) & 1


def bits_to_bytes(bits, byte_count: int) -> bytes:
    output = bytearray()
    current = 0
    for index, bit in enumerate(bits, start=1):
        current = (current << 1) | int(bit)
        if index % 8 == 0:
            output.append(current)
            current = 0
            if len(output) == byte_count:
                break
    if len(output) != byte_count:
        raise NoHiddenDataError("The hidden payload is incomplete.")
    return bytes(output)


def encode_b64(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).decode("ascii")


def decode_b64(data: str) -> bytes:
    return base64.urlsafe_b64decode(data.encode("ascii"))

File: server/src/services/steganography.ts
import crypto from "node:crypto";
import path from "node:path";
import bmp from "bmp-js";
import { PNG } from "pngjs";

const MAGIC = Buffer.from("SIHS1", "ascii");
const FORMAT_VERSION = 1;
const KDF_ITERATIONS = 390_000;
const SUPPORTED_EXTENSIONS = new Set([".png", ".bmp"]);

export class SteganographyError extends Error {}
export class UnsupportedImageError extends SteganographyError {}
export class CapacityError extends SteganographyError {}
export class NoHiddenDataError extends SteganographyError {}
export class PasswordRequiredError extends SteganographyError {}
export class InvalidPasswordError extends SteganographyError {}

export interface HiddenPayload {
  kind: "text" | "file";
  data: Buffer;
  filename: string | null;
  contentType: string;
  encrypted: boolean;
}

interface RgbaImage {
  width: number;
  height: number;
  data: Buffer;
}

export function embedSecretInCover(input: {
  coverBytes: Buffer;
  coverFilename: string;
  secretData: Buffer;
  payloadKind: "text" | "file";
  payloadFilename?: string | null;
  contentType?: string | null;
  password?: string | null;
}) {
  const packageBytes = buildPackage({
    secretData: input.secretData,
    payloadKind: input.payloadKind,
    payloadFilename: input.payloadFilename ?? null,
    contentType: input.contentType ?? "application/octet-stream",
    password: input.password ?? null
  });
  const image = decodeSupportedImage(input.coverBytes, input.coverFilename);
  return embedPackage(image, packageBytes);
}

export function extractSecretFromImage(stegoBytes: Buffer, password?: string | null): HiddenPayload {
  const image = decodeSupportedImage(stegoBytes, "stego.png", false);
  const packageBytes = extractPackage(image);
  return parsePackage(packageBytes, password ?? null);
}

export function buildPackage(input: {
  secretData: Buffer;
  payloadKind: "text" | "file";
  payloadFilename: string | null;
  contentType: string;
  password: string | null;
}) {
  if (!input.secretData.length) {
    throw new SteganographyError("Secret payload cannot be empty.");
  }

  const encrypted = Boolean(input.password);
  let payloadBytes = input.secretData;
  const metadata: Record<string, unknown> = {
    version: FORMAT_VERSION,
    kind: input.payloadKind,
    filename: input.payloadFilename,
    content_type: input.contentType,
    encrypted
  };

  if (encrypted) {
    const salt = crypto.randomBytes(16);
    const nonce = crypto.randomBytes(12);
    const key = deriveKey(input.password || "", salt, KDF_ITERATIONS);
    const cipher = crypto.createCipheriv("aes-256-gcm", key, nonce);
    cipher.setAAD(MAGIC);
    payloadBytes = Buffer.concat([cipher.update(input.secretData), cipher.final()]);
    const authTag = cipher.getAuthTag();
    metadata.kdf = {
      name: "PBKDF2HMAC-SHA256",
      iterations: KDF_ITERATIONS,
      salt: salt.toString("base64url")
    };
    metadata.nonce = nonce.toString("base64url");
    metadata.auth_tag = authTag.toString("base64url");
  }

  const header = Buffer.from(JSON.stringify(metadata), "utf8");
  const headerLength = Buffer.alloc(4);
  headerLength.writeUInt32BE(header.length, 0);
  return Buffer.concat([MAGIC, headerLength, header, payloadBytes]);
}

export function parsePackage(packageBytes: Buffer, password: string | null): HiddenPayload {
  if (!packageBytes.subarray(0, MAGIC.length).equals(MAGIC) || packageBytes.length < MAGIC.length + 4) {
    throw new NoHiddenDataError("No hidden SIHS payload was found.");
  }

  const headerLength = packageBytes.readUInt32BE(MAGIC.length);
  const headerStart = MAGIC.length + 4;
  const headerEnd = headerStart + headerLength;
  if (headerLength <= 0 || headerEnd > packageBytes.length) {
    throw new NoHiddenDataError("The hidden payload metadata is invalid.");
  }

  let metadata: Record<string, unknown>;
  try {
    metadata = JSON.parse(packageBytes.subarray(headerStart, headerEnd).toString("utf8"));
  } catch (error) {
    throw new NoHiddenDataError("The hidden payload metadata is invalid.");
  }

  if (metadata.version !== FORMAT_VERSION || (metadata.kind !== "text" && metadata.kind !== "file")) {
    throw new NoHiddenDataError("The hidden payload version or type is not supported.");
  }

  let payloadBytes = packageBytes.subarray(headerEnd);
  const encrypted = Boolean(metadata.encrypted);
  if (encrypted) {
    if (!password) {
      throw new PasswordRequiredError("A password is required to extract this payload.");
    }
    try {
      const kdf = metadata.kdf as { salt: string; iterations: number } | undefined;
      const salt = Buffer.from(String(kdf?.salt ?? ""), "base64url");
      const iterations = Number(kdf?.iterations ?? 0);
      const nonce = Buffer.from(String(metadata.nonce ?? ""), "base64url");
      const authTag = Buffer.from(String(metadata.auth_tag ?? ""), "base64url");
      const key = deriveKey(password, salt, iterations);
      const decipher = crypto.createDecipheriv("aes-256-gcm", key, nonce);
      decipher.setAAD(MAGIC);
      decipher.setAuthTag(authTag);
      payloadBytes = Buffer.concat([decipher.update(payloadBytes), decipher.final()]);
    } catch (error) {
      throw new InvalidPasswordError("The provided password is incorrect.");
    }
  }

  return {
    kind: metadata.kind,
    data: payloadBytes,
    filename: metadata.filename ? String(metadata.filename) : null,
    contentType: String(metadata.content_type ?? "application/octet-stream"),
    encrypted
  };
}

function decodeSupportedImage(bytes: Buffer, filename: string, checkExtension = true): RgbaImage {
  const extension = path.extname(filename).toLowerCase();
  if (checkExtension && !SUPPORTED_EXTENSIONS.has(extension)) {
    throw new UnsupportedImageError("Only PNG and BMP images are supported.");
  }

  if (!checkExtension || extension === ".png") {
    try {
      const png = PNG.sync.read(bytes);
      return { width: png.width, height: png.height, data: Buffer.from(png.data) };
    } catch (error) {
      if (checkExtension) {
        throw new UnsupportedImageError("The uploaded file is not a readable PNG image.");
      }
    }
  }

  if (extension === ".bmp") {
    try {
      const decoded = bmp.decode(bytes);
      return { width: decoded.width, height: decoded.height, data: Buffer.from(decoded.data) };
    } catch (error) {
      throw new UnsupportedImageError("The uploaded file is not a readable BMP image.");
    }
  }

  throw new UnsupportedImageError("Only PNG and BMP images are supported.");
}

function embedPackage(image: RgbaImage, packageBytes: Buffer) {
  const frame = Buffer.alloc(4 + packageBytes.length);
  frame.writeUInt32BE(packageBytes.length, 0);
  packageBytes.copy(frame, 4);
  const capacityBits = image.width * image.height * 3;
  const requiredBits = frame.length * 8;
  if (requiredBits > capacityBits) {
    throw new CapacityError(
      `Secret payload needs ${frame.length} bytes, but this image can hold ${Math.floor(capacityBits / 8)} bytes.`
    );
  }

  let bitIndex = 0;
  for (const byte of frame) {
    for (let shift = 7; shift >= 0; shift -= 1) {
      const bit = (byte >> shift) & 1;
      const pixelIndex = Math.floor(bitIndex / 3) * 4;
      const channel = bitIndex % 3;
      image.data[pixelIndex + channel] = (image.data[pixelIndex + channel] & 0xfe) | bit;
      bitIndex += 1;
    }
  }

  const png = new PNG({ width: image.width, height: image.height, inputColorType: 6 });
  image.data.copy(png.data);
  return PNG.sync.write(png);
}

function extractPackage(image: RgbaImage) {
  if (image.width * image.height * 3 < 32) {
    throw new NoHiddenDataError("No hidden SIHS payload was found.");
  }

  const packageLengthBuffer = readBits(image, 0, 32);
  const packageLength = packageLengthBuffer.readUInt32BE(0);
  const capacity = Math.floor((image.width * image.height * 3) / 8) - 4;
  if (packageLength <= 0 || packageLength > capacity) {
    throw new NoHiddenDataError("No hidden SIHS payload was found.");
  }

  const packageBytes = readBits(image, 32, packageLength * 8);
  if (!packageBytes.subarray(0, MAGIC.length).equals(MAGIC)) {
    throw new NoHiddenDataError("No hidden SIHS payload was found.");
  }
  return packageBytes;
}

function readBits(image: RgbaImage, startBit: number, bitCount: number) {
  const output = Buffer.alloc(Math.ceil(bitCount / 8));
  for (let i = 0; i < bitCount; i += 1) {
    const bitIndex = startBit + i;
    const pixelIndex = Math.floor(bitIndex / 3) * 4;
    const channel = bitIndex % 3;
    const bit = image.data[pixelIndex + channel] & 1;
    output[Math.floor(i / 8)] = (output[Math.floor(i / 8)] << 1) | bit;
  }
  return output;
}

function deriveKey(password: string, salt: Buffer, iterations: number) {
  if (!salt.length || !iterations) {
    throw new InvalidPasswordError("The provided password is incorrect.");
  }
  return crypto.pbkdf2Sync(password, salt, iterations, 32, "sha256");
}
